mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -138,24 +138,59 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       // Limit the use of hashDist since it's controversial
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
-
-      // Compute threshold to get around k elements.
-      // To guarantee to have enough neighbors in one pass, we need (p - err) * N >= M
-      // so we pick quantile p = M / N + err
-      // M: the number of nearest neighbors; N: the number of elements in dataset
-      val relativeError = 0.05
-      val approxQuantile = numNearestNeighbors.toDouble / count + relativeError
       val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
-      if (approxQuantile >= 1) {
-        modelDatasetWithDist
+
+      val spark = dataset.sparkSession
+      import spark.implicits._
+
+      if (numNearestNeighbors < 1000) {
+        val r = Random.nextInt
+        val distColIdx = modelDatasetWithDist.schema.fieldNames.indexOf(distCol)
+        val rows = modelDatasetWithDist
+          .rdd
+          .map { row =>
+            val dist = row.getDouble(distColIdx)
+            (r, (dist, row))
+          }.aggregateByKey(new BoundedPriorityQueue[(Double, Row)](numNearestNeighbors)(
+            Ordering.by[(Double, Row), Double](_._1).reverse))(
+            seqOp = (c, v) => c += v,
+            combOp = (c1, c2) => c1 ++= c2
+          ).flatMap { case (_, c) => c.iterator.map(_._2) }
+        spark.createDataFrame(rows, modelDatasetWithDist.schema)
+
       } else {
-        val hashThreshold = modelDatasetWithDist.stat
-          .approxQuantile(distCol, Array(approxQuantile), relativeError)
-        // Filter the dataset where the hash value is less than the threshold.
-        modelDatasetWithDist.filter(hashDistCol <= hashThreshold(0))
+        val relativeError = 0.05
+        val (summaries, count) = modelDatasetWithDist
```

Comment:
Here we compute QuantileSummaries & count together, it should be faster than existing impl.